crypto/txscript/src/lib.rs

```diff
@@ -0,0 +1,489 @@
+        let pubkey_script = utxo_entry.script_public_key.script();
+        // The pubkey in P2SH is just validating the hash on the OpMultiSig script
+        // the user provides
+        let is_p2sh = (pubkey_script.len() == 35) && // 3 opcodes number + 32 data
+                (pubkey_script[0] == opcodes::codes::OpBlake2b) &&
+                (pubkey_script[1] == opcodes::codes::OpData32) &&
+                (pubkey_script[pubkey_script.len() -1] == opcodes::codes::OpEqual);
+        match id < tx.tx().inputs.len() {
+            true => Ok(Self {
+                dstack: Default::default(),
+                astack: Default::default(),
+                script_source: ScriptSource::TxInput { tx, input, id, utxo_entry, is_p2sh },
+                reused_values,
+                sig_cache,
+                cond_stack: Default::default(),
+                num_ops: 0,
+            }),
+            false => Err(TxScriptError::InvalidIndex(id, tx.tx().inputs.len())),
+        }
+    }
+
+    pub fn from_script(
+        script: &'a [u8],
+        reused_values: &'a mut SigHashReusedValues,
+        sig_cache: &'a Cache<SigCacheKey, Result<(), secp256k1::Error>>,
+    ) -> Self {
+        Self {
+            dstack: Default::default(),
+            astack: Default::default(),
+            script_source: ScriptSource::StandAloneScripts(vec![script]),
+            reused_values,
+            sig_cache,
+            cond_stack: Default::default(),
+            num_ops: 0,
+        }
+    }
+
+    #[inline]
+    pub fn is_executing(&self) -> bool {
+        // TODO: check values
+        return self.cond_stack.is_empty() || *self.cond_stack.first().expect("Checked not empty") != 1;
+    }
+
+    fn execute_opcode(&mut self, opcode: Box<dyn OpCodeImplementation<T>>) -> Result<(), TxScriptError> {
+        // Different from kaspad: Illegal and disabled opcode are checked on execute instead
+        // Note that this includes OP_RESERVED which counts as a push operation.
+        if opcode.value() > NO_COST_OPCODE {
+            self.num_ops += 1;
+            if self.num_ops > MAX_OPS_PER_SCRIPT {
+                return Err(TxScriptError::TooManyOperations(MAX_OPS_PER_SCRIPT));
+            }
+        } else if opcode.len() > MAX_SCRIPT_ELEMENT_SIZE {
+            return Err(TxScriptError::ElementTooBig(opcode.len(), MAX_SCRIPT_ELEMENT_SIZE));
+        }
+
+        if self.is_executing() || opcode.is_conditional() {
+            if opcode.value() > 0 && opcode.value() <= 0x4e {
+                opcode.check_minimal_data_push()?;
+            }
+            opcode.execute(self)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn execute_script(&mut self, script: &[u8]) -> Result<(), TxScriptError> {
+        let script_result = script
+            .iter()
+            .batching(|it| {
+                // reads the opcode num item here and then match to opcode
+                it.next().map(|code| deserialize(*code, it))
```

Comment:
I think it'll be clearer if `deserialize` will be renamed to `deserialize_opcode_data` and the current lambda will have its own `deserialize_opcode` function